app/src/main/java/org/simple/clinic/patientcontact/PatientContactEvent.kt

```diff
@@ -13,3 +17,15 @@ data class PatientProfileLoaded(
 data class OverdueAppointmentLoaded(
     val overdueAppointment: Optional<OverdueAppointment>
 ) : PatientContactEvent()
+
+data class NormalCallClicked(
+    override var permission: Optional<RuntimePermissionResult> = None,
+    override val permissionRequestCode: Int = 1,
+    override val permissionString: String = Manifest.permission.CALL_PHONE
```

Comment:
Few questions here:
1. Are we going to be passing these values from somewhere else apart from the `permission` itself?
2. Why in `NormalCallClicked` do we need to pass a different `permissionString` than `CALL_PHONE`?
3. Why did we choose to set default values here rather than passing them from the relevant `Update` class?

Reply:
Ah, this is because of the way we switched to doing runtime permissions requesting.

I found the older way of requesting permissions was leaking into business logic, and I wanted to move it out of this and manage it before it goes into the Mobius loop.

As a side-effect of doing this, I had to implement this in such a way that these events have these properties that have to be set in the constructor. Please see [this PR](https://github.com/simpledotorg/simple-android/pull/1243) for the relevant details.

I also think it's possible to do this via an annotation, but I haven't gotten time to explore this yet.

Reply:
The tl;dr is that these properties are required to be defined by any type that implements the `RequiresPermission` interface since it uses them to do the automatic permissions management.

I can also convert them to functions instead in another PR if you think that might be more explicit.

Reply:
I understand now.
Although the thing that I want to point here is that while we are moving runtime permissions from business logic, we are adding them to `Event` instead and eventually this looks like we are using `Event` to do more than what it's expected to. I don't think it is a problem if the business logic knows it needs permission to execute something. This could very well have been a side effect as well. Wdyt?
Also, why did we choose to make `RequiresPermission` an `interface` and not a `data class` which can be passed around in an event? You can have default values there as well but they can be set in the `data class` itself instead of `Event`.

Reply:
I know that this is already implemented this way now so I don't expect it to be changed.
I am just trying to make some things clear for myself.

Reply:
> I understand now.
> Although the thing that I want to point here is that while we are moving runtime permissions from business logic, we are adding them to `Event` instead and eventually this looks like we are using `Event` to do more than what it's expected to. I don't think it is a problem if the business logic knows it needs permission to execute something. This could very well have been a side effect as well. Wdyt?
> Also, why did we choose to make `RequiresPermission` an `interface` and not a `data class` which can be passed around in an event? You can have default values there as well but they can be set in the `data class` itself instead of `Event`.

The reason is because it needs to be a type. If you look at the `RequestPermissions` observable transformer, this looks at incoming events to see which ones require permissions (which is does by examining the type) and then holds them in-flight while the permission request flow is done and then forwards them again with the request result. If this is a property on the event, then I would have to check for the presence of this property using reflection instead.

Reply:
One thing which I can think of for the future is still have the `RequiresPermission` interface:

```kotlin
fun interface RequiresPermission {
  var grant: Optional<PermissionResult> = None
}
```

and then change the definition of which permissions are required to an annotation instead:

```kotlin
@Permission(
  value = Manifest.permission.CALL_PHONE,
  requestCode = 100
)
object NormalCall: ContactPatientEvent, RequiresPermission
```

I am not sure if this would work out but I can try it out if you feel it looks better.

Reply:
> I understand now.
> Although the thing that I want to point here is that while we are moving runtime permissions from business logic, we are adding them to `Event` instead and eventually this looks like we are using `Event` to do more than what it's expected to. I don't think it is a problem if the business logic knows it needs permission to execute something. This could very well have been a side effect as well. Wdyt?
> Also, why did we choose to make `RequiresPermission` an `interface` and not a `data class` which can be passed around in an event? You can have default values there as well but they can be set in the `data class` itself instead of `Event`.

In response to your other concern, since permissions are something that are specific to the Android platform, I did not want this logic of how to request for permissions themselves to the Mobius loop. It is sufficient for our use case that when the mobius loop receives the event, it knows whether permission was granted or not for the event to proceed.

What do you think?

Reply:
Sure. This looks better. Let's discuss this then.